language/move-lang/src/expansion/translate.rs

```diff
@@ -661,6 +670,70 @@ fn struct_fields(
     E::StructFields::Defined(field_map)
 }
 
+//**************************************************************************************************
+// Friends
+//**************************************************************************************************
+
+fn friend(
+    context: &mut Context,
+    friends: &mut UniqueMap<ModuleIdent, E::Friend>,
+    pfriend: P::Friend,
+) {
+    let (mident, friend) = friend_(context, pfriend);
+    match &friend.value {
+        E::Friend_::Ok => {}
+        E::Friend_::UnresolvedError => {
+            assert!(context.has_errors());
+        }
+    }
+    if let Some(old_friend) = friends.get(&mident) {
+        let msg = format!(
+            "Duplicate friend declaration '{}'. Friend declarations in a module must be unique",
+            mident
+        );
+        context.error(vec![
+            (friend.loc, msg),
+            (old_friend.loc, "Previously declared here".into()),
+        ])
+    }
+    if let Err(_old_loc) = friends.add(mident, friend) {
+        assert!(context.has_errors());
+    };
+}
+
+fn friend_(context: &mut Context, sp!(loc, pfriend): P::Friend) -> (ModuleIdent, E::Friend) {
+    assert!(context.exp_specs.is_empty());
+    let (mident, efriend) = match pfriend {
+        P::Friend_::Module(mname) => match context.aliases.module_alias_get(&mname.0).cloned() {
+            None => {
+                context.error(vec![(
+                    mname.loc(),
+                    format!("Unbound module alias '{}'", mname),
+                )]);
+                (
+                    ModuleIdent(sp(
+                        mname.loc(),
+                        ModuleIdent_ {
+                            name: mname,
+                            address: context.cur_address(),
+                        },
+                    )),
```

Comment:
Assign a temp local here to make this easier to read

Reply:
This part of code is removed as unresolved aliases are simply dropped from the table now.